corporate/lib/stripe.py

```diff
@@ -1745,7 +1745,16 @@ def process_initial_upgrade(
                 customer=customer, next_invoice_date=next_invoice_date, **plan_params
             )
 
+            self.write_to_audit_log(
+                event_type=AuditLogEventType.CUSTOMER_PLAN_CREATED,
+                event_time=event_time,
+                extra_data=plan_params,
+            )
+
             if plan.status < CustomerPlan.LIVE_STATUS_THRESHOLD:
+                # Tier and usage limit change will happen when plan becomes live.
+                self.do_change_plan_type(tier=plan_tier)
```

Comment:
Is this reordering just for better error handling, in case a failure happens after this? The commit message should really explain why the refactor makes sense.

I'm also not 100% sure why the plan type should change at a separate time from `tier`? That kinda seems like a bug to me.

Reply:
Sorry, I didn't mention it but the main reason for this is to call `do_change_plan_type` before `self.invoice_plan(plan, event_time)` below.

Reply:
OK, but can you spell out why that's helpful?

Reply:
I still don't understand -- shouldn't we only be using properties of the `CustomerPlan`, not the `RemoteZulipServer`, in the billing process?

Reply:
oh yeah, we recently changed it in #29222, but do you see any downsides to this change. I can image if we add a check related to current status of `Realm` in `invoice_plan` in future and having it reflect the correct tier of plan which is synced with the active `CustomerPlan` would make that check valid.

Reply:
I think it's probably safe, just wanted to make sure I know what's happening.

Reply:
Yeah this seems fine, it is right next to where we set `tier` on the plan itself.